### PR TITLE
modules: canopennode: add rxmsg callback

### PR DESCRIPTION
**Pull request by chgabriel79**
Implement callback for incoming CAN messages for any of the configured filters for CANopenNode.

This can be used to wake the loop calling CO_process when a new message was received which needs processing.

### CODE REVIEW

File: modules/canopennode/canopennode.h
Comment (henrikbrixandersen) [inline modules/canopennode/canopennode.h:168]:
```suggestion
 * @param callback the callback to set
```

File: modules/canopennode/CO_driver.c
Comment (henrikbrixandersen) [inline modules/canopennode/CO_driver.c:115]:
```suggestion
			if (rxmsg_callback != NULL) {
```

File: modules/canopennode/CO_driver.c
Comment (henrikbrixandersen) [inline modules/canopennode/CO_driver.c:34]:
```suggestion
static canopen_rxmsg_callback rxmsg_callback = NULL;
```

File: modules/canopennode/CO_driver.c
Comment (henrikbrixandersen) [inline modules/canopennode/CO_driver.c:116]: This suffers from TOCTOU (https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use), but I guess the idea is to install this callback once and never "uninstall" it by setting it to `NULL`?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good, thanks. A few minor comments.

File: modules/canopennode/CO_driver.c
Comment (chgabriel79) [reply modules/canopennode/CO_driver.c]: You're right, that was the idea. But I changed it anyway, it's safer and more consistent with how callbacks are handled in the CAN drivers.

**Comment by chgabriel79** [comment]
I added your suggestions, and also clarified that the callback is run in interrupt context (CAN rx callback). And I had some problems with my client not adding the files to the commit, sorry for the noise...

File: modules/canopennode/canopennode.h
Comment (henrikbrixandersen) [inline modules/canopennode/canopennode.h:155]: Thanks! One last thing that I didn't notice before:

```suggestion
typedef void (*canopen_rxmsg_callback_t)(void);
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by alexanderwachter — DISMISSED** [review]
Looks good.
Just a thought:
Is a callback disable function necessary?
Or maybe in the API description that a NULL ptr disables the callback?

File: modules/canopennode/CO_driver.c
Comment (chgabriel79) [reply modules/canopennode/CO_driver.c]: The compliance check does not seem to like this change. Shall I revert it?

**Comment by chgabriel79** [comment]
> Or maybe in the API description that a NULL ptr disables the callback?

It states "Setting a new callback will replace any existing callback."

I used a similar wording to other callbacks, e.g. `can_set_state_change_callback` which states "Calling this function again overrides any previously registered callback." (and also does not explicitly mention passing NULL disables the callback).

File: modules/canopennode/CO_driver.c
Comment (henrikbrixandersen) [reply modules/canopennode/CO_driver.c]: Yes, sorry about that.

**Review by alexanderwachter — APPROVED** [review]

**Review by martinjaeger — APPROVED** [review]

**Comment by kartben** [comment]
Commenting to see if that's enough to get the PR to show as green in the merge dashboard 🤷🏼

**Comment by kartben** [comment]
@henrikbrixandersen do you want to try rebasing in case (unlikely IMO) merge dashboard is right by saying something is wrong with CI somehow?
Alternatively, just do a merge batch and pick this PR since AFAICT it *is* green

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen do you want to try rebasing in case (unlikely IMO) merge dashboard is right by saying something is wrong with CI somehow? Alternatively, just do a merge batch and pick this PR since AFAICT it _is_ green

It was missing my approval after the latest force-push, so no - the merge-list had it correctly listed.

**Pull request merged**
